dbms/src/DataTypes/SmallestJSON/SmallestJSONSerialization.cpp

```diff
@@ -0,0 +1,357 @@
+               NumberImpl(field_column, ext::bit_cast<UInt64>(value));
+    }
+
+    bool StringImpl(IColumn * column, const char * data, rapidjson::SizeType length)
+    {
+        if (ColumnString * column_string = static_cast<ColumnString *>(column))
+        {
+            ColumnString::Chars & column_string_data = column_string->getChars();
+            ColumnString::Offsets & column_offsets_data = column_string->getOffsets();
+
+            if (!column_offsets_data.size() && row_size)
+            {
+                column_string_data.resize_fill(row_size, 0);
+                for (size_t index = 0; index < row_size; ++index)
+                    column_offsets_data.push_back(index + 1);
+            }
+
+            column_string_data.insert(data, data + length);
+            column_string_data.push_back(0);
+            column_offsets_data.push_back(column_string_data.size());
+            return true;
+        }
+        else
+            throw Exception("It is bug", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    bool String(const char * data, rapidjson::SizeType length, bool /*copy*/)
+    {
+        IColumn * mark_column = current_struct_info->getOrCreateMarkColumn();
+        IColumn * field_column = current_struct_info->getOrCreateDataColumn(STRING_DATA_TYPE);
+
+        return NumberImpl<true>(mark_column, UInt8(TypeIndex::String), maybeFillingJSONTypeToMark(current_struct_info)) &&
+               StringImpl(field_column, data, length);
+    }
+
+    bool Key(const char * data, rapidjson::SizeType length, bool /*copy*/)
+    {
+        current_struct_info = parent_struct_info->getOrCreateChildren(StringRef(data, length));
+        return true;
+    }
+
+    bool StartObject()
+    {
+        parent_struct_info = current_struct_info;
+
+        if (current_struct_info->mark_column)
+            NumberImpl<true>(current_struct_info->mark_column, UInt8(TypeIndex::SmallestJSON));
+
+        return true;
+    }
+
+    bool EndObject(rapidjson::SizeType /*type*/)
+    {
+        parent_struct_info = parent_struct_info->getParent();
+        current_struct_info = current_struct_info->getParent();
+        return true;
+    }
+
+    bool RawNumber(const Ch * /*ch*/, rapidjson::SizeType /*type*/, bool /*b*/)
+    {
+        throw Exception("Method RawNumber is not supported for JSONStructAndColumnBinder.", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    bool StartArray()
+    {
+        throw Exception("SmallestJSON type does not support parse array json.", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    bool EndArray(rapidjson::SizeType /*type*/)
+    {
+        throw Exception("SmallestJSON type does not support parse array json.", ErrorCodes::NOT_IMPLEMENTED);
+    }
+};
+
+
+template <typename InputStream>
+void SmallestJSONSerialization::deserialize(DB::IColumn & column_, const DB::FormatSettings & settings, InputStream & input_stream)
+{
+    rapidjson::Reader reader;
```

Comment:
`simdjson` shows better performance than `rapidjson` in our tests. I suppose there could be an option what to use. You can add a template parameter and pass `RapidJSONParser` or `SimdJSONParser` to implement that if you like.